### PR TITLE
Adds windup autofire functionality. The future is now. (Later)

### PR DESCRIPTION
**Pull request by necromanceranne**
## About The Pull Request

Pulls the wind up autofire functionality from my energy weapon pr as a standalone code improvement. The original use case for this was for the Deltra, which you can see [here](https://github.com/tgstation/tgstation/pull/67768). That's not in this PR, it's just a working example.

## Why It's Good For The Game

I think it'd be real cool to use this for something. That is to say, I know **I** want to use it for something, but why wait on me?

## Changelog

:cl:
code: Added code functionality for autofire windup and resetting, both redefinable. Nothing uses this code just yet!
/:cl:

### CODE REVIEW

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/datums/components/fullauto.dm [hunk 1]
```diff
@@ -11,16 +11,28 @@
 	var/autofire_shot_delay = 0.3 SECONDS //Time between individual shots.
 	var/mouse_status = AUTOFIRE_MOUSEUP //This seems hacky but there can be two MouseDown() without a MouseUp() in between if the user holds click and uses alt+tab, printscreen or similar.
 
+	//windup autofire vars
+	var/windup_autofire = FALSE //Whether the delay between shots increases over time, simulating a spooling weapon. Resets when you stop firing.
+	var/current_windup_reduction = 0 //the reduction to shot delay for windup. Resets when you stop firing to 0.
+	var/windup_autofire_reduction_multiplier = 0.3 //the percentage of autfire_shot_delay that is added to current_windup_reduction. In the default example, every shot reduces the shot delay by 30%.
+	var/windup_autofire_cap = 0.3 //How high of a reduction that current_windup_reduction can reach. In the default example, the delay between shots would be 30% of the original fire delay.
+	var/windup_spindown = 3 SECONDS //How long it takes for weapons that have spooled-up to reset back to the original firing speed
```
[the remark this answers is not in the document]
Comment (necromanceranne) [reply code/datums/components/fullauto.dm:19]: Sure thing

File: code/datums/components/fullauto.dm [hunk 2]
```diff
@@ -11,16 +11,28 @@
...
+	var/timerid
 	COOLDOWN_DECLARE(next_shot_cd)
 
-/datum/component/automatic_fire/Initialize(_autofire_shot_delay)
+/datum/component/automatic_fire/Initialize(_autofire_shot_delay, _windup_autofire, _windup_autofire_reduction_multiplier, _windup_autofire_cap, _windup_spindown)
```
[the remark this answers is not in the document]
Comment (necromanceranne) [reply code/datums/components/fullauto.dm:23]: Right, was just being consistent with what was there.

File: code/datums/components/fullauto.dm [hunk 3]
```diff
@@ -241,6 +257,11 @@
 	stop_autofiring()
 	return FALSE
 
+/datum/component/automatic_fire/proc/windup_reset(deltimer)
```
[the remark this answers is not in the document]
Comment (necromanceranne) [reply code/datums/components/fullauto.dm:260]: Sure.

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/datums/components/fullauto.dm [hunk 4]
```diff
@@ -241,6 +262,12 @@
 	stop_autofiring()
 	return FALSE
 
+// Reset for our windup, resetting everything back to initial values after a variable set amount of time (determined by var/windup_spindown).
+/datum/component/automatic_fire/proc/windup_reset(deltimer)
```
[the remark this answers is not in the document]
Comment (necromanceranne) [reply code/datums/components/fullauto.dm:266]: Oops I just missed one /

**Pull request merged**